File: python_datastructs/LRU.py
class Node():
    def __init__(self, k, v):
        self.key = k
        self.val = v


class LRUCache():

    def __init__(self, capacity):
        self.capacity = capacity
        self.mapping = {}
        head = None
        tail = None

    def get(self, key):
        if key in self.mapping:
            n = self.mapping[key]
            self.remove(n)
            self.insertFront(n)
            return n.val
        else:
            return None

    def put(self, key, val):
        if key in self.mapping:
            node = self.mapping[key]
            node.val = val
            self.remove(node)
            self.insertFront(node)

        else:
            node = Node(key, val)
            if len(self.mapping) > capacity:
                self.remove(self.tail)
                del self.mapping[self.tail.key]
                self.insertFront(node)
            else:
                self.insertFront(node)

    def remove(self, node):
        if node.prevNode:
            node.prevNode.nextNode = node.nextNode
        else:
            self.head = node.nextNode

        if node.nextNode:
            node.nextNode.prevNode = node.prevNode
        else:
            self.tail = node.prevNode

    def insertFront(self, node):
        if not self.head:
            self.head = node
            self.tail = node

        else:
            node.nextNode = self.head
            self.head = node

    def printCache(self, rev=False):
        node = self.head if not rev else self.tail
        while (node):
            print(node.val)
            node = node.nextNode if not rev else node.prevNode

lru = LRUCache(2)
a = Node('a', 23)
b = Node('b', 51)
c = Node('c', 58)
d = Node('d', 68)
e = Node('e', 76)
lru.put(a)
lru.printCache()
lru.put(b)
lru.printCache()
lru.put(c)
lru.printCache()
lru.put(d)
lru.printCache()
lru.put(e)
lru.printCache()

File: java_datastructs/LRU.java
import java.util.HashMap;

// front = MRU
// back = LRU

class LRU {
    private HashMap<String, Node> dict;
    private Node head;
    private Node tail;
    private int capacity;

    class Node {
        private int value;
        private Node nextNode;
        private Node prevNode;

        public Node(int value) {
            this.value = value;
            this.nextNode = null;
            this.prevNode = null;
        }

        public void setValue(int value) {
            this.value = value;
        }

        public int getValue() {
            return this.value;
        }
    }

    public LRU(int capacity) {
        this.capacity = capacity;
        this.dict = new HashMap<String, Node>();
        this.head = null;
        this.tail = null;
    }

    private Node getNode(String key) {
        return this.dict.get(key);
    }

    private int getVal(String key) {
        Node node = getNode(key);
        if (node != null) {
            return node.getValue();
        } else { // correct behaviour?
            return 0;
        }
    }

    private void put(String key, int value) {
        if (this.dict.containsKey(key)) {
            Node oldNode = this.dict.get(key);
            if (oldNode != null) {
                oldNode.setValue(value);
                removeFromList(oldNode);
                addFront(oldNode);
            } else {
                Node newNode = new Node(value);
                this.dict.put(key, newNode);
                addFront(newNode);
            }

        } else { // must create a new Node
            Node newNode = new Node(value);
            if (this.dict.size() < this.capacity) {
                // add to the front
                this.dict.put(key, newNode);
                addFront(newNode);
            } else { // need to remove the LRU before adding a new one
                evictLRU();
                addFront(newNode);
            }
        }
    }

    private Node evictLRU() {

    }
}
